File: app/src/main/java/com/lumiyaviewer/lumiya/react/RequestForwarder.java
package com.lumiyaviewer.lumiya.react;

import com.lumiyaviewer.lumiya.react.-$Lambda$swF2K5wPKI2_xA-bWP-XwHVnywU.AnonymousClass1;
import com.lumiyaviewer.lumiya.react.Subscription.OnData;
import com.lumiyaviewer.lumiya.react.Subscription.OnError;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executor;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

public abstract class RequestForwarder<Kup, Tup, Kdown, Tdown> implements RequestHandler<Kup> {
    @Nullable
    private final Executor executor;
    private final Object lock = new Object();
    @Nonnull
    private final ResultHandler<Kup, Tup> resultHandler;
    @Nonnull
    private final Subscribable<Kdown, Tdown> subscribable;
    private final Map<Kup, DownstreamSubscription> subscriptions = new HashMap();

    private class DownstreamSubscription implements OnData<Tdown>, OnError, UnsubscribableOne {
        private final Kup key;
        private final Subscription<Kdown, Tdown> subscription;

        private DownstreamSubscription(Kup kup, Kdown kdown) {
            this.key = kup;
            this.subscription = RequestForwarder.this.subscribable.subscribe(kdown, RequestForwarder.this.executor, this, this);
        }

        /* synthetic */ DownstreamSubscription(RequestForwarder requestForwarder, Object obj, Object obj2, DownstreamSubscription downstreamSubscription) {
            this(obj, obj2);
        }

        /* access modifiers changed from: package-private */
        /* renamed from: lambda$-com_lumiyaviewer_lumiya_react_RequestForwarder$DownstreamSubscription_827  reason: not valid java name */
        public /* synthetic */ void m30lambda$com_lumiyaviewer_lumiya_react_RequestForwarder$DownstreamSubscription_827(Object obj) {
            RequestForwarder.this.processResultInternal(this.key, obj);
        }

        public void onData(Tdown tdown) {
            if (RequestForwarder.this.executor != null) {
                RequestForwarder.this.executor.execute(new -$Lambda$swF2K5wPKI2_xA-bWP-XwHVnywU(this, tdown));
            } else {
                RequestForwarder.this.processResultInternal(this.key, tdown);
            }
        }

        public void onError(Throwable th) {
            RequestForwarder.this.resultHandler.onResultError(this.key, th);
        }

        public void unsubscribe() {
            this.subscription.unsubscribe();
        }
    }

    public RequestForwarder(@Nonnull RequestSource<Kup, Tup> requestSource, @Nonnull Subscribable<Kdown, Tdown> subscribable, @Nullable Executor executor) {
        this.executor = executor;
        this.subscribable = subscribable;
        this.resultHandler = requestSource.attachRequestHandler(this);
    }

    private void processRequestInternal(@Nonnull Kup kup) {
        DownstreamSubscription downstreamSubscription = new DownstreamSubscription(this, kup, getDownstreamKey(kup), null);
        synchronized (this.lock) {
            downstreamSubscription = (DownstreamSubscription) this.subscriptions.put(kup, downstreamSubscription);
        }
        if (downstreamSubscription != null) {
            downstreamSubscription.unsubscribe();
        }
    }

    private void processResultInternal(@Nonnull Kup kup, @Nullable Tdown tdown) {
        this.resultHandler.onResultData(kup, processResult(tdown));
    }

    protected abstract Kdown getDownstreamKey(@Nonnull Kup kup);

    /* synthetic */ void handleRequestProcessing(Object obj) {
        processRequestInternal(obj);
    }

    public void onRequest(@Nonnull Kup kup) {
        if (this.executor != null) {
            this.executor.execute(new AnonymousClass1(this, kup));
        } else {
            processRequestInternal(kup);
        }
    }

    public void onRequestCancelled(@Nonnull Kup kup) {
        DownstreamSubscription downstreamSubscription;
        synchronized (this.lock) {
            downstreamSubscription = (DownstreamSubscription) this.subscriptions.remove(kup);
        }
        if (downstreamSubscription != null) {
            downstreamSubscription.unsubscribe();
        }
    }

    protected abstract Tup processResult(@Nullable Tdown tdown);
}

File: app/src/main/kotlin/com/lumiyaviewer/lumiya/react/RequestForwarder.kt
package com.lumiyaviewer.lumiya.react

import com.lumiyaviewer.lumiya.react.Subscription.OnData
import com.lumiyaviewer.lumiya.react.Subscription.OnError
import java.util.HashMap
import java.util.concurrent.Executor

abstract class RequestForwarder<KUp, TUp, KDown, TDown>(
    requestSource: RequestSource<KUp, TUp>,
    private val subscribable: Subscribable<KDown, TDown>,
    private val executor: Executor?
) : RequestHandler<KUp> {

    private val lock = Any()
    private val resultHandler: ResultHandler<KUp, TUp> = requestSource.attachRequestHandler(this)
    private val subscriptions = HashMap<KUp, DownstreamSubscription>()

    private inner class DownstreamSubscription(
        private val key: KUp,
        downstreamKey: KDown
    ) : OnData<TDown>, OnError, UnsubscribableOne {
        
        private val subscription: Subscription<KDown, TDown> = 
            subscribable.subscribe(downstreamKey, executor, this, this)

        override fun onData(data: TDown) {
            if (executor != null) {
                executor.execute {
                    processResultInternal(key, data)
                }
            } else {
                processResultInternal(key, data)
            }
        }

        override fun onError(error: Throwable) {
            resultHandler.onResultError(key, error)
        }

        override fun unsubscribe() {
            subscription.unsubscribe()
        }
    }

    protected abstract fun getDownstreamKey(upstreamKey: KUp): KDown
    
    protected abstract fun processResult(downstreamData: TDown?): TUp

    private fun processRequestInternal(key: KUp) {
        val newSubscription = DownstreamSubscription(key, getDownstreamKey(key))
        val oldSubscription = synchronized(lock) {
            subscriptions.put(key, newSubscription)
        }
        oldSubscription?.unsubscribe()
    }

    private fun processResultInternal(key: KUp, data: TDown?) {
        resultHandler.onResultData(key, processResult(data))
    }

    override fun onRequest(key: KUp) {
        if (executor != null) {
            executor.execute {
                processRequestInternal(key)
            }
        } else {
            processRequestInternal(key)
        }
    }

    override fun onRequestCancelled(key: KUp) {
        val subscription = synchronized(lock) {
            subscriptions.remove(key)
        }
        subscription?.unsubscribe()
    }
}
